packages/api/src/api/project/post.ts

```diff
@@ -0,0 +1,36 @@
+import { Schema } from '@hangar/shared';
+import { Request, Response } from 'express';
+import { logger } from '../../utils/logger';
+import { validatePayload } from '../../utils/validatePayload';
+import { Project, User } from '@hangar/database';
+
+export const post = async (req: Request, res: Response) => {
+  const { entityManager: em, user } = req;
+
+  const { errorHandled, data } = validatePayload({ req, res, schema: Schema.project.post });
+  if (errorHandled) return;
+  
+  const query = {id:req.user.id};
+  
+  // check for existing project
+  try {
+    const user = await em.findOne(User, query);
+    if (user?.project!==undefined) {
+      res.status(409);
+      return;
+    }
+  }
+  catch {
+    res.status(500);
+    return;
+  }
+  try {
+    const project = new Project(data);
+    em.persistAndFlush(project);
```

Comment:
These two lines are the ones that need to be wrapped in a transaction; you basically need to do the following:
- start a transaction
- use **_that_** entity manager to re-fetch the user but in doing so, use a `LockMode.PESSIMISTIC_WRITE_OR_FAIL` locking mode (check me on that but I'm pretty sure that's what we want)
  - If that call fails, the transaction will fail and you'll end up in your catch block; see my comment below about how to handle that
- If the user was obtainable, create the project, persist the project, then update the user with the project, persist the user, flush both

Reply:
[]()****@ ****

Reply:
@